sdk/src/asset_handlers/woff_io.rs

```diff
@@ -0,0 +1,1739 @@
+    //   if our goal is simply to describe the provenance _of an already-existing asset_,
+    //   then perhaps we need to permit incorrectly-ordered SFNTs/WOFF1s? In that
+    //   case, then perhaps the selection of BTreeMap versus Vec is done at
+    //   construction time, ensuring the desired behavior?
+    // - Otherwise, we could just use BTreeMap for SFNT/WOFF1 and Vec for WOFF2
+    // - Other matters?
+    tables: BTreeMap<TableTag, Table>,
+}
+
+impl Font {
+    /// New blank font
+    pub fn new(magic: Magic) -> Self {
+        Self {
+            magic,
+            tables: BTreeMap::new(),
+        }
+    }
+
+    /// Reads in a font file.
+    fn read<T: Read + Seek + ?Sized>(source_stream: &mut T) -> core::result::Result<Font, Error> {
+        // Must always rewind input
+        source_stream.rewind()?;
+        // Verify the font has a valid version in it before assuming the rest is
+        // valid (NOTE: we don't actually do anything with it, just as a safety check).
+        let font_magic_u32: u32 = source_stream.read_u32::<BigEndian>()?;
+        let font_magic: Magic = <u32 as std::convert::TryInto<Magic>>::try_into(font_magic_u32)
+            .map_err(|_err| Error::UnsupportedFontError)?;
+        // Check the magic number
+        match font_magic {
+            Magic::OpenType|Magic::TrueType => todo!("Yow! Implement read_sfnt!"),
+            Magic::Woff => Font::read_woff(source_stream),
+            Magic::Woff2 => todo!("Yow! Implement read_woff2!"),
+        }
+    }
+
+    /// Writes out a font file.
+    fn write<TDest: Write + ?Sized>(&mut self, destination: &mut TDest) -> Result<()> {
+        // Find the header - we must have one
+        let woff_hdr = match self.tables.get_mut(&WOFF_HEADER_TAG) {
+            Some(Table::WoffHeader(woff_hdr)) => {
+                // We have a WOFF header, so we can write it out
+                woff_hdr
+            },
+            _ => {
+                // We don't have a WOFF header, so we can't write it out
+                Err(Error::FontSaveError)?
+            }
+        };
+
+        //TBD - is a rewind appropriate here?
+        //destination.rewind()?;
+
+        // Write the header
+        woff_hdr.write(destination)?;
+
+        // Write the table directory. This needs work; we need some way to
+        // preserve the physical order implied by the offsets/sizes
+        // of the font as we read it in.
+        //
+        // Options:
+        // - Explicitly store the Table Directory as a pseudo-table of some kind
+        // - Add the offset information to our font table structure. For example,
+        //   the Table enum could change from being a struct-enum to a tuple
+        //   enum which includes the offset and the table-struct.
+        // - Other ideas?
+        for (tag, table) in &mut self.tables {
+            match tag {
+                &WOFF_HEADER_TAG|&WOFF_METADATA_TAG|&WOFF_PRIVATE_DATA_TAG|
+                &TABLE_DIRECTORY_TAG => {
```

Comment:
I'm not sure where/when we write the table directory, am I missing something?

Reply:
Yeah, there is a dilemma as to whether it's worth storing the Table Directory as a chunk, since (a) you can synthesize it pretty easily, and (b) we are likely to muck with it, at least a little bit.